### ISSUE TITLE
Proposal: Global Repository Permissions Caching

### ISSUE THREAD

**Issue by tsenart** [sourcegraph/sourcegraph#4812]
## Abstract

This proposal outlines a potential new approach to repository permissions caching.

## Background

At a high level, the current implementation of repository permissions verification works like this:

- Load repositories from Postgres.
- If the authenticated user is an admin, return all repos.
- Load all external user accounts associated with the authenticated user from Postgres.
- For each configured authorization provider:
  - If none of the loaded external accounts match this provider:
    - Fetch an external account from this provider (i.e. via `FetchAccount`)
    - Associate this external account with the authenticated user and persist it to Postgres.
  - Filter out repositories that don't belong to this provider (based on `api.ExternalRepoSpec`).
  - Load repository permissions for the selected repositories (via `RepoPerms`).
    - Return cached permissions if there's an unexpired cache entry for this user.
    - Otherwise, fill the cache with fresh permissions from the provider's code host and return those.
  - Add those repositories for which the current user has the required permissions to the `accepted` set.
  - If there are still repositories to verify, loop to the next provider.
- Return accepted repositories in the same order as the input repos slice.

#### Cache filling slows down interactive user requests

Since cache filling happens synchronously in the user request path, it makes the user wait for as long as it takes to load all the requested permissions. How often this happens depends on the configured cache expiration `ttl`. Admins who are concerned with the inflicted load on their code hosts can increase this setting to a large duration at the expense of freshness of data (i.e. changes in permissions in the code host will take up to `ttl` time to be reflected on Sourcegraph).

This problem manifests when either the code host has slow-downs or when loading permissions for a large number of repos is just slow, which is the case for some of our largest customers (e.g. loading permissions for **~10k** repositories from the Bitbucket Server API takes roughly **10s**).

#### Missing cache filling concurrency control

Neither the GitHub or the GitLab authorization providers control the concurrency of cache fills for a single user. If users happen to send multiple requests when their permissions cache have expired, all or many of these requests will concurrently refresh the permissions cache from the code host.

This is implemented in the Bitbucket Server provider, but it's not in an ideal state. The pessimistic locking approach used there, leveraging Postgres row locking, has the potential to exhaust the Postgres connection pool quickly, when faced with higher concurrency than the configured pool size.

#### Maintainability burden from lack of consistency and duplication

Each authorization provider is re-implementing many cross cutting concerns, chief among them, permissions caching. There are other redundancies that can be eliminated, such as the duplicated `Repos` method that partitions repositories into `mine` and `theirs` whose implementations simply call out to `authz.GetCodeHostRepos`.

This duplication of efforts opens the doors to bugs and makes it harder to maintain this sub-system. Moreover, caching repository permissions outside of each provider, at the `authzFilter` level, would result in a significant performance boost.

## Proposal

To address the problems described above, we propose to _unify repository permissions caching_ outside of `authz.Provider` implementations. With a central place that handles this concern, we can address all of the identified problems at once, for all providers.

1. **Async cache filling**: Instead of blocking a user's request we'd use the stale cache immediatelly and start a refresh in the background (i.e. in a go-routine).
2. **Concurrency control**: We'd start controlling concurrency of cache fills for every provider. Instead of Postgres row-level locking, we'd use an alternative method that isn't prone to exhaust the connection pool (e.g optimisitically update the `updated_at` timestamp the row first, so that other processes don't concurrently cache fill, and then update the other columns when the data becomes available). Additional concurrency control of cache fills of all users would be done via well-tuned `bitbucketserver.Client` rate limiting.
3. **Maintainability**: An `Authorizer` type would be introduced, with its logic based on `authzFilter`, but using dependency injection for easier testing. This type would handle all caching of permissions while each `authz.Provider` would only take care of fetching fresh permissions.
4. **Performance**: For most user requests, since caching of permissions is now global instead of per provider, filtering repos via the `Authorizer` would be a simple in-memory set intersection using the authenticated user's repository permissions cache, represented as `roaring.Bitmap` of authorized repository IDs.

## Rationale

This proposal kills many birds with one stone. However, it is a significant change to a security sensitive sub-system. The risk of introducing bugs can be managed by splitting changes in manageable chunks that will be thoroughly reviewed and tested.

## Compatibility

- Individual `ttl` settings on each authorization provider could be replaced by a global setting that'd apply to all configured providers. If we pursure this route, deprecation of these settings would be necessary.
- Redis ought to be cleaned up after this change since it holds significant amounts of data related to permissions caching of GitHub and GitLab providers.

## Implementation

This work fits in with the 3.6 milestone goal of the Core Services team since it would decrease the 99th percentile latency of searches significantly. It's possible that it will transpose to the next milestone for its full completion, but here's a proposed implementation order that could bring some of the benefits to users already in the 3.6 release.

- [x] Change the current Bitbucket Server provider's `store` to use an alternative to row-level locking. By itself, this would prevent potential connection pool exhaustion in the scenarios described before.
- [x] Change the current Bitbucket Server provider to update its caches asynchronously. This would allow us to recommend Bitbucket Server ACLs to some of our largest customers after we release 3.6, with caveats (see _Open Issues_).
- [ ] Building on the previous work, lift the `store` outside of the `bitbucket` provider package, to the top-level authz package as `PermsStore`.
- [ ] Introduce the `Authorizer` type in the `authz` package. This type would use an injected `PermsStore` for all its caching. Implement the equivalent of `authzFilter` as a method in this type: `func (a *Authorizer) AuthorizedRepos(ctx context.Context, user *types.User, repos []*types.Repo) ([]*types.Repo, error)`.
- [ ] Reuse the `Authorizer` in the code-paths where `authzFilter` is currently used.
- [ ] Change the `GitHub` authorization providers to not do any caching themselves.
- [ ] Change the `GitLab` authorization providers to not do any caching themselves.

### Cold caches

This proposal doesn't address the _cold cache_ issue. When new users are created, or caches are wiped, we can't use a stale cache in user requests because there is none. In this scenario, users will have to wait. This is particularly problematic when Sourcegraph is deployed anew, since potentially many users can storm it and all of them would have to wait: not a great first experience.

Here are two complementary ideas that could alleviate this issue:

1. Document this clearly and ask admins to not open the flood-gates all at once.
2. When creating a new user, start a permissions cache fill in the background. This raises the chances that when the user hits a page that needs authorization, the permissions will already be there.

**Comment by tsenart** [comment]
@beyang, @sourcegraph/core-services: Any feedback appreciated!

@christinaforney, @nicksnyder, @sqs: Feedback on implementation plan, schedule and open issues appreciated.

**Comment by nicksnyder** [comment]
Thanks for writing out such a clear proposal! It makes sense to me and I think this work can be prioritized. Does this proposal require the full @sourcegraph/core-services team, or would @keegancsmith and/or @mrnugget be able to continue working on other perf improvements?

> Redis ought to be cleaned up after this change since it holds significant amounts of data related to permissions caching of GitHub and GitLab providers.

Does this need to happen explicitly or would this data silently/automatically expire?

**Comment by beyang** [comment]
A few initial thoughts:
* With an async approach, what sort of SLA can we provide on latency between a permissions update and when that update is reflected in Sourcegraph? Some customers might want very short latency here. The need for low latency is not just when a new user is created. It's also when any permission is updated on the code host.
* We might be doing more work than is necessary currently. E.g., we shouldn't need to compute the entire matrix of user-repo permissions, and doing so might be computationally burdensome even in an async manner. In theory, we should only need to compute permissions for the set of a repositories that is being displayed to an end user at a given point in time.
* There is no standard permissions API across code hosts and our manner of computing permissions varies depending on the code host (and we even support multiple methods for a single code host depending). We currently cache data that is very close to what the code host returns. This is efficient and straightforward. If we switch to caching data that is more "derived", it could lead to an increase in the size of the things we need to cache. What's the back-of-the-envelope here on how much more we need to cache?

**Comment by tsenart** [comment]
> Does this proposal require the full @sourcegraph/core-services team, or would @keegancsmith and/or @mrnugget be able to continue working on other perf improvements?

We can certainly share some of the implementation steps, if necessary. We'll touch base next week to assess our priorities.

> Does this need to happen explicitly or would this data silently/automatically expire?

I think all permissions caching data of GitHub and GitLab providers is currently stored in redis-cache with TTLs. If that's the case, we won't have to do anything. But I'll double check.

**Comment by tsenart** [comment]
> With an async approach, what sort of SLA can we provide on latency between a permissions update and when that update is reflected in Sourcegraph? 

The same as with the sync approach, but we don't make the user wait by using stale permissions while the update is happening in a background go-routine. So the upper bound on the permissions update would be `TTL + duration(cache-fill)`.

**Comment by tsenart** [comment]
> We might be doing more work than is necessary currently. E.g., we shouldn't need to compute the entire matrix of user-repo permissions, and doing so might be computationally burdensome even in an async manner. In theory, we should only need to compute permissions for the set of a repositories that is being displayed to an end user at a given point in time.

Thanks for reminding me of this (we talked about it in our 1-1). I'll explore the performance implications of doing authorization **after** search results are returned, but I suspect that'll be tricky, specially with pagination, and might not be worth it. As it currently stands, we have to run authorization on all repositories at once, for the search use case.

> There is no standard permissions API across code hosts and our manner of computing permissions varies depending on the code host (and we even support multiple methods for a single code host depending). We currently cache data that is very close to what the code host returns. This is efficient and straightforward. If we switch to caching data that is more "derived", it could lead to an increase in the size of the things we need to cache. What's the back-of-the-envelope here on how much more we need to cache?

We will need to cache less, not more. As we do in the Bitbucket Server provider today, we'd cache a highly compressed bit set of all the Sourcegraph repository IDs a given user has permissions for. 

For a hypothetical upper bound of 100k repositories, and 10k users, the aggregate memory usage would be rounding 329,36 MB per frontend API instance. For 30k repositories and 1k users, that number would be 16,472MB. This is computed with https://godoc.org/github.com/RoaringBitmap/roaring#BoundSerializedSizeInBytes

**Comment by keegancsmith** [comment]
Proposal looks great. Given we have a few eyes on this proposal, I'd like to revisit the TTL model. It implies that until a TTL expires, a users global set of repositories is static. IE if a repo is added / deleted / hidden it won't be reflected for a subset of users. Admins will always have fresh data since they do not get auth checks. I can imagine in general this model is surprising. However, I don't know of a better model which is performant and doesn't require some very very coupled and fragile code mimicking the auth model of a code host. Any thoughts?

>> We might be doing more work than is necessary currently. E.g., we shouldn't need to compute the entire matrix of user-repo permissions, and doing so might be computationally burdensome even in an async manner. In theory, we should only need to compute permissions for the set of a repositories that is being displayed to an end user at a given point in time.
>
> Thanks for reminding me of this (we talked about it in our 1-1). I'll explore the performance implications of doing authorization after search results are returned, but I suspect that'll be tricky, specially with pagination, and might not be worth it. As it currently stands, we have to run authorization on all repositories at once, for the search use case.

I think switching to a post-filtering model still suffers from a lot of the same issues, and new ones. All the work we did in repo-updater syncing was essentially trying to get us away from post-filtering due to the poor performance and expense it had on code hosts.

Additionally here are some other reasons against this idea:
- not enforcing this at the db package layer seems like a way to accidently leak private data
- search can return a lot of data for repos you can't show? Do we just keep paginating + post-filtering until we find results? This would be very poor performance as well.
- Most of our value add is "global"/cross repo related. We need to ensure whatever model we have is cheap when we want to look at potentially large sets of repos.

**Comment by nicksnyder** [comment]
> So the upper bound on the permissions update would be `TTL + duration(cache-fill)`.

This isn't entirely true unless the cache fill happens automatically when the TTL expires. My understanding of the current proposal is to lazily fill the cache on the next user request. This means that as long as the stale cache entry persists, the user can make one request using the old permissions an unbounded amount of time later (which seems undesirable).

Here is an idea:
- TTL is an absolute upper bound on how long permissions can be cached. When permissions expire due to TTL, permissions must be refetched synchronously during a search.
- Permissions are considered stale after TTL/2 seconds. The goal is to refresh permissions _before_ they expire. A stale permission is still used for searches, but is refreshed asynchronously on _any_ page visit by a given user. This keeps the cache warm for active users, respects the configured TTL, and hides the latency from searches.

> IE if a repo is added / deleted / hidden it won't be reflected for a subset of users

If a repo is deleted/hidden it shouldn't matter if it is in the permissions cache because there won't be any results anyway, right?

As soon as we detect that the repo set has changed, we could mark the permission cache as stale (as if TTL/2 has elapsed) so that the next page visit by the user causes permissions to be refreshed.

**Comment by tsenart** [comment]
> This means that as long as the stale cache entry persists, the user can make one request using the old permissions an unbounded amount of time later (which seems undesirable).

Ah, yes, good catch.

> TTL is an absolute upper bound on how long permissions can be cached. When permissions expire due to TTL, permissions must be refetched synchronously during a search.

I think we'd need at a minimum some "loading" UI to show users that permissions are being fetched. If our hard TTL is, say, 6h, when everyone comes back to work the next day, we'll have a thundering herd against the code host, and, if we have rate limiting in place, each individual cache refresh can actually take a long time.

The alternative to this is to have a background process, like the repo updater git scheduler, than continually keeps permissions caches up to date, given the configured TTL, managing its own schedule.

> Permissions are considered stale after TTL/2 seconds. The goal is to refresh permissions before they expire. A stale permission is still used for searches, but is refreshed asynchronously on any page visit by a given user. This keeps the cache warm for active users, respects the configured TTL, and hides the latency from searches.

This is a good idea if we manage to take care of the thundering herd issue mentioned above. Is a TTL of 24h acceptable to admins? That would prevent hitting the hard TTL in that scenarion and allow this to be a solution. But I don't think 24h is gonna fly.

> As soon as we detect that the repo set has changed, we could mark the permission cache as stale (as if TTL/2 has elapsed) so that the next page visit by the user causes permissions to be refreshed.

Actually, we could detect that some repos we loaded from the database aren't in the permissions cache and load permissions for just those repos, then adding them to the cache. The TTL would remain intact.

**Comment by nicksnyder** [comment]
> Is a TTL of 24h acceptable to admins?

A good question that I don't know the answer to. Can you directly ask customers that will be using this feature?

**Comment by tsenart** [comment]
> A good question that I don't know the answer to. Can you directly ask customers that will be using this feature?

Yes, we can definitely do that. But the more I think about it, the less it seems like the right design, given these TTL constraints. It seems like the background process approach is superior from a user experience perspective. @keegancsmith: Thoughts?

**Comment by keegancsmith** [comment]
The background process could be a decent solution, but I worry about the load. Additionally I'd make the hard TTL something like 3 days (for weekends), and the soft one whatever people have for the TTL at the moment. I would love to find out of admins find that acceptable, since that would be the easiest solution.

**Comment by tsenart** [comment]
@dadlerj, @christinaforney: What would be the best way to get this information from interested customers? We'd like to know if a hard TTL of user permissions of three days would be acceptable.

**Comment by dadlerj** [comment]
@slimsag can ask https://app.hubspot.com/contacts/2762526/company/407948923

I'll ask https://app.hubspot.com/contacts/2762526/company/557889642

**Comment by tsenart** [comment]
@dadlerj, @slimsag: Be sure to explain the trade-off to the customer when asking this question. The trade-off is: The lower the TTL, the higher the load on the code host and the worse user experience (since the user has to wait more often for their permissions to be refreshed). Loading permissions for a single user is essentially loading all repos for which the given user has permissions for. Code hosts don't offer a more efficient way to fetch this information at this point, unfortunately.

**Comment by slimsag** [comment]
(I reached out to them and will let you know when I hear back, most likely that'll be late next week or while we're at Gophercon)

**Comment by christinaforney** [comment]
The user experience (as described by Stephen):
1. If a repository exists on Sourcegraph and a user _has_ access to it, you then _revoke their access_ on Bitbucket, they would continue to have access on Sourcegraph for _up to_ three days (max).
2. If a repository exists on Sourcegraph and a user _doesn't_ have access to it, you then _grant them permission_ on Bitbucket, they would only have access to that repository on Sourcegraph after 1.5 days (if the repository is new or not known by Sourcegraph, they would get this access instantly.)

Thoughts/questions:
- Could we write a manual override in Sourcegraph for an admin to “sync access” for a particular user? The UX would be something along the lines of insert a username and force syncing of that person’s permissions to override the cache OR a repository or project name that we would force syncing of permissions for.
- Alternatively could we write a webhook as part of the Bitbucket Server plugin (or something else) that would push information from Bitbucket Server that an ACL has changed, and could notify Sourcegraph of this information to trigger updates to the appropriate users/cache?

**Comment by nicksnyder** [comment]
> If a repository exists on Sourcegraph and a user has access to it, you then revoke their access on Bitbucket, they would continue to have access on Sourcegraph for up to three days (max).

To be more precise, they would continue to have access to Sourcegraph for up to 1.5 days plus one more query up to another 1.5 days after that. Hard to communicate the nuance there though.

BTW, my proposal was `soft ttl = hard ttl / 2`, but we could choose another constant like `hard ttl / 4` which would reduce the 1.5 day number to 0.75 day (at the expense of more queries to the code host).

**Comment by tsenart** [comment]
> Could we write a manual override in Sourcegraph for an admin to “sync access” for a particular user? The UX would be something along the lines of insert a username and force syncing of that person’s permissions to override the cache OR a repository or project name that we would force syncing of permissions for.

Yes, we could! It'd be nice if this would be a good enough control for admins. Let's see what the feedback and proceed from there.

> Alternatively could we write a webhook as part of the Bitbucket Server plugin (or something else) that would push information from Bitbucket Server that an ACL has changed, and could notify Sourcegraph of this information to trigger updates to the appropriate users/cache?

Here be dragons. There are no web-hooks of permissions change events. There's an audit-log API, but we'd have to essentially replicate the permissions logic on our end by inferring how changes to group membership, project membership, individual repository permissions, and more, would affect the permissions of a single user. It's too risky, and a lot of effort, so it's better to treat the code host as the source of truth. Additionally, we need a solution that works for all code hosts uniformly.

**Comment by tsenart** [comment]
After #5400 is merged, https://app.hubspot.com/contacts/2762526/company/407948923 will be able to try out the solution we have agreed to implement, even though we have learned it isn't a requirement for going GA.

The remainder of this issue's planned work will be put on hold as I will begin work on A8N.

cc @slimsag @christinaforney @sqs @nicksnyder @sourcegraph/core-services

**Comment by beyang** [comment]
Related RFC proposal: https://docs.google.com/document/d/167WV_rhBg3oj0yw9fHwUMlB-8BR3ktk2FSsKiUQo55Y/edit

**Comment by beyang** [comment]
@tsenart Can we close this in favor of the [RFC]( https://docs.google.com/document/d/167WV_rhBg3oj0yw9fHwUMlB-8BR3ktk2FSsKiUQo55Y/edit)?

**Comment by tsenart** [comment]
@unknwon: Can you capture the follow up work that you took from this proposal that is left to do in separate issues? (i.e. using the `authz.Store` with the existing providers). Once that's done we can close this I believe.

**Comment by unknwon** [comment]
> @unknwon: Can you capture the follow up work that you took from this proposal that is left to do in separate issues? (i.e. using the `authz.Store` with the existing providers). Once that's done we can close this I believe.

I created https://github.com/sourcegraph/sourcegraph/issues/6845.
